x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClientCache.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.amazonbedrock;
+
+import com.amazonaws.http.IdleConnectionReaper;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyMap;
+
+public final class AmazonBedrockInferenceClientCache implements AmazonBedrockClientCache {
+
+    private final BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator;
+    private volatile Map<Integer, AmazonBedrockBaseClient> clientsCache = emptyMap();
+
+    public AmazonBedrockInferenceClientCache(BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator) {
+        this.creator = creator;
+    }
+
+    public AmazonBedrockBaseClient getOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        var returnClient = internalGetOrCreateClient(model, timeout);
+        flushExpiredClients();
+        return returnClient;
+    }
+
+    private AmazonBedrockBaseClient internalGetOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
```

Comment:
My understanding of this method is that if multiple threads retrieve an already existing client, an internal reference will be incremented for each retrieval. I haven't looked through all the code yet but do we call `close()` on the client?

I believe that'd mean we'd continue to increment the reference but only decrement when the client expires?

Reply:
Yep - all the clients _should_ be wrapped in a `try(var client = cache.getOrCreateClient()) {` block which will auto-close it when it goes out of scope... I'll double check though

Reply:
I did find one reference (in the executor) that doesn't properly auto-close this (fixed now in my local code)... the current setup should keep the reference at 1 (unless it's actively in use by the executor), and then when expired, it should set to 0 and release it now.

Reply:
(note - I've updated my code to remove the reference counting)